src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs

```diff
@@ -303,33 +298,28 @@ private DecisionTree AddByType(DecisionTree.ByType byType, TypeSymbol type, Deci
                     }
                 }
             }
-            foreach (var kvp in byType.TypeAndDecision)
+
+            // if the last type is the type we need, add to it
+            DecisionTree result = null;
+            if (byType.TypeAndDecision.Count != 0)
             {
-                var MatchedType = kvp.Key;
-                var Decision = kvp.Value;
-                // See if matching Type matches this value
-                switch (ExpressionOfTypeMatchesPatternType(type, MatchedType, ref _useSiteDiagnostics))
+                var lastTypeAndDecision = byType.TypeAndDecision.Last();
+                if (lastTypeAndDecision.Key.TupleUnderlyingTypeOrSelf() == type.TupleUnderlyingTypeOrSelf())
```

Comment:
It is impossible to use tuple types (with names) as the type of a declaration pattern. There simply isn't any syntax for it. So it cannot appear as a type in the map. Moreover, whether or not those are taken into account can only make a difference for performance, not correctness (until we support recursive patterns).

I'll file an issue to investigate, test, and fix if necessary the behavior for nested tuple types.

As for `dynamic`, it is not intended to be supported as the type of a declaration pattern (no runtime object ever has the type `dynamic`). However, I note the compiler does accept it; I'll file a bug.

---
In reply to: [100591895](https://github.com/dotnet/roslyn/pull/16928#discussion_r100591895) [](ancestors = 100591895)